### ISSUE TITLE
Relax calendar/fixing requirement for CSA compounding rate in NettedExposureCalculator::collateralPaths

### ISSUE THREAD

**Issue by rkapl123** [OpenSourceRisk/Engine#99]
Dear All!

As we also need to calculate xva values on good friday, I'd like to suggest a relaxation on the date requirement resp. the exact calculation of the CSA compounding rate in `NettedExposureCalculator::collateralPaths`:

```cpp
    // Don't use Settings::instance().evaluationDate() here, this has moved to simulation end date.
    Date today = market_->asofDate();
    string csaIndexName = netting->csaDetails()->index();
    // this line throws an Exception if called on good-friday or any other holiday of the resp. CSA index:
    Real csaRateToday = market_->iborIndex(csaIndexName, configuration_)->fixing(today);
    LOG("CSA compounding rate for index " << csaIndexName << " = " << csaRateToday);
```

A possibility might be to use the `forecastTodaysFixing` parameter, however then still the `QL_REQUIRE` would need to be moved after checking this parameter (and returning the forecast rate) in `InterestRateIndex::fixing`:
```cpp
Rate InterestRateIndex::fixing(const Date& fixingDate,
                                   bool forecastTodaysFixing) const {

        QL_REQUIRE(isValidFixingDate(fixingDate),
                   "Fixing date " << fixingDate << " is not valid");

        Date today = Settings::instance().evaluationDate();

        if (fixingDate>today ||
            (fixingDate==today && forecastTodaysFixing))
            return forecastFixing(fixingDate);
```

Another, less intrusive alternative would be to introduce a separate method `fixingEstimate` here, that is only used for the collateralPaths method above and does the above mentioned relaxation, avoiding the removal of the fixing date check for future forecast fixing dates in general.

I'd like to make a pull request after comments/suggestions on that.

-regards,
Roland

**Comment by pcaspers** [comment]
Dear Roland,

my suggestion would be to adjust the fixing date to the last valid date before "todays", i.e. do something like

```
Real csaRateToday = market_->iborIndex(csaIndexName, configuration_)->
        fixing(market_->iborIndex(csaIndexName, configuration_)->fixingCalendar().adjust(today, P));
```

I think this is close to what actually happens in reality, where the compounding over a weekend or over holidays is done using the last available fixing before that holiday.

Also, this kind of adjustment is what we do at some other places at the code as well.

What do you think?

Best Regards
Peter

**Comment by rkapl123** [comment]
Dear Peter!

Yes, that sounds very good. I'll do the PR accordingly.

regards,
Roland

**Comment by rkapl123** [comment]
I've included the corrections in PR , commit #[6e8fb01](https://github.com/OpenSourceRisk/Engine/pull/98/commits/6e8fb010aa02ba8f03e118a58d4669098ac4db41)

**Comment by QuaternionRisk** [comment]
OK, then I'll close the issue.
